### ISSUE TITLE
SIM401 triggers on dict like objects that aren't proper dictionaries

### ISSUE THREAD

**Issue by jogo-openai** [astral-sh/ruff#15814]
### Description

```
$ ruff --version
ruff 0.9.3
```

```python
class Foo:
    def __init__(self):
        self._dict = {}

    def __getitem__(self, key):
        return self._dict[key]

    def __setitem__(self, key, value):
        self._dict[key] = value

    def __iter__(self):
        return self._dict.__iter__()



foo = Foo()
foo['key'] = 'value'


if "key" in foo:
    value = foo["key"]
else:
    value = None

print(value)


# print(foo.get('key', None))
# Triggers `AttributeError: 'Foo' object has no attribute 'get'`
```

`ruff check --fix  --unsafe-fixes --diff --select=SIM401  foo.py`
```
--- foo.py
+++ foo.py
@@ -17,10 +17,7 @@
 foo['key'] = 'value'


-if "key" in foo:
-    value = foo["key"]
-else:
-    value = None
+value = foo.get('key', None)

 print(value)


Would fix 1 error.
```

**Comment by dylwil3** [comment]
I'm slightly in favor of just updating the documentation (which is part of #15584) for this unsafe fix. An alternative is to restrict to instances where our limited type inference can tell that we have an honest `dict` object, but that would be a significant restriction in applicability. That said, false positives are more annoying than false negatives for this sort of rule, so I could be convinced of the second route.

**Comment by jogo-openai** [comment]
I have a slight preference for removing false positives and making sure things are actually dicts. That being said,  just updating the documentation for this unsafe fix also seems good to me (I suspect this means llarge code bases would opt to not use this rule).

**Comment by MichaReiser** [comment]
SIM401 was added in 2023 when we didn't had any type inference in Ruff, which is why we didn't restrict it to *known dicts*. We've since then started to restrict new rules to known dicts only (e.g. RUF051) and I think it makes sense to do the same here. 

This is technically a breaking change for users that suppressed `SIM401` in cases where the value isn't a true dict but removing the false positives seems like a clear improvement.

**Comment by junhsonjb** [comment]
Hi! I'm super excited to be able to participate in this awesome project! I'm looking forward to working with you all towards a fix on this! Let me know if there's anything I need to change/fix/add/etc. 

I really like ruff so thanks for all y'all do on this project and for the community! Cheers!
